src/libraries/Microsoft.PowerFx.Interpreter/Functions/LibraryMutation.cs

```diff
@@ -0,0 +1,279 @@
+                }
+                else
+                {
+                    result = await tableValue.PatchAsync(baseRecord.Value, updatesRecord.Value, cancellationToken).ConfigureAwait(false);
+                }
+
+                if (result.IsError)
+                {
+                    return result.ToFormulaValue();
+                }
+
+                resultRows.Add(result);
+            }
+
+            return new InMemoryTableValue(IRContext.NotInSource(tableValue.Type), resultRows);
+        }
+    }
+
+    // If arg1 is pure PFx record, it will return a runtime internal error.
+    // Patch(DS, table_of_rows_with_updates)
+    internal class PatchAggregateSingleTableImpl : PatchAggregateSingleTableFunction, IAsyncTexlFunction5
+    {
+        public async Task<FormulaValue> InvokeAsync(IServiceProvider runtimeServiceProvider, FormulaType irContext, FormulaValue[] args, CancellationToken cancellationToken)
+        {
+            MutationProcessUtils.ThrowIfPFxV1NotActive(runtimeServiceProvider.GetService<Features>(), "Patch");
+
+            var arg0 = args[0];
+
+            if (args[0] is LambdaFormulaValue arg0lazy)
+            {
+                arg0 = await arg0lazy.EvalAsync().ConfigureAwait(false);
+            }
+
+            if (arg0 is not TableValue tableValue)
+            {
+                return arg0;
+            }
+
+            if (args[1] is not TableValue tableUpdates)
+            {
+                return args[1];
+            }
+
+            List<DValue<RecordValue>> resultRows = new List<DValue<RecordValue>>();
+
+            foreach (var row in tableUpdates.Rows)
+            {
+                cancellationToken.ThrowIfCancellationRequested();
+
+                if (row.IsError)
+                {
+                    return row.ToFormulaValue();
+                }
+
+                if (row.IsBlank)
+                {
+                    continue;
+                }
+
+                var result = await tableValue.PatchAsync(row.Value, cancellationToken).ConfigureAwait(false);
+
+                if (result.IsError)
+                {
+                    return result.ToFormulaValue();
+                }
+
+                resultRows.Add(result);
+            }
+
+            return new InMemoryTableValue(IRContext.NotInSource(tableValue.Type), resultRows);
+        }
+    }
+
+    // Patch(Record, Updates1, Updates2,…)
+    internal class PatchRecordImpl : PatchRecordFunction, IAsyncTexlFunction5
+    {
+        public async Task<FormulaValue> InvokeAsync(IServiceProvider runtimeServiceProvider, FormulaType irContext, FormulaValue[] args, CancellationToken cancellationToken)
+        {
+            MutationProcessUtils.ThrowIfPFxV1NotActive(runtimeServiceProvider.GetService<Features>(), "Patch");
+
+            return MutationProcessUtils.MergeRecords(args).ToFormulaValue();
+        }
```

Comment:
This is not a mutation function, it's a pure one (no side-effects, self-contained). The implementation can go to the core library.

Reply:
Although I agree this could be moved, in terms of code maintainability its better to leave it this way IMO. Patch is a mutation function (with this  overload as exception).